### PR TITLE
Add AWS ResourceTag to HostedCluster and NodePool

### PR DESCRIPTION
**Pull request by alvaroaleman**
This change adds AWS ResourceTags into the HostedCluster and Nodepool
apis. HostedCluster tags will be set on  capi AWS cluster and NodePool
tags will be merged with HostedCluster tags where the latter tags
precendence and then set on the capi MachineTemplate.

Ref https://issues.redhat.com/browse/HOSTEDCP-148

/cc @derekwaynecarr @csrwng

### CODE REVIEW

File: api/v1alpha1/hostedcluster_types.go [hunk 1]
```diff
@@ -324,6 +324,34 @@ type AWSPlatformSpec struct {
...
 	// an AWS credentials file.
 	NodePoolManagementCreds corev1.LocalObjectReference `json:"nodePoolManagementCreds"`
+
+	// resourceTags is a list of additional tags to apply to AWS resources created for the cluster.
+	// See https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html for information on tagging AWS resources.
+	// AWS supports a maximum of 50 tags per resource. OpenShift reserves 25 tags for its use, leaving 25 tags
+	// available for the user.
+	// +kubebuilder:validation:MaxItems=25
+	// +optional
+	ResourceTags []AWSResourceTag `json:"resourceTags,omitempty"`
```
Comment (alvaroaleman) [inline api/v1alpha1/hostedcluster_types.go:334]: The format is copied from the openshift api types and different from the capi types, where this is a typed `map[string]string`: https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L361-L381

Not sure which is better and no strong opinion either way.

File: api/v1alpha1/nodepool_types.go [hunk 2]
```diff
@@ -185,6 +185,15 @@ type AWSNodePoolPlatform struct {
...
 	RootVolume *Volume `json:"rootVolume,omitempty"`
+
+	// resourceTags is a list of additional tags to apply to AWS nodes.
+	// These will be merged with Cluster-level tags and Cluster-level tags take precedence in case of conflicts.
+	// See https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html for information on tagging AWS resources.
+	// AWS supports a maximum of 50 tags per resource. OpenShift reserves 25 tags for its use, leaving 25 tags
+	// available for the user.
+	// +kubebuilder:validation:MaxItems=25
+	// +optional
+	ResourceTags []AWSResourceTag `json:"resourceTags,omitempty"`
```
Comment (alvaroaleman) [inline api/v1alpha1/nodepool_types.go:196]: Should the two ResourceTags be immutable? Can capi reconcile them on existing resources?

File: api/v1alpha1/hostedcluster_types.go (hunk 1 above)
Comment (enxebre) [reply api/v1alpha1/hostedcluster_types.go:334]: From a UX pov the one you propose seems fine and clear to me and matches ocp. From a technical pov the one you propose I think is probably easier and more granular to validate via kubebuilder tags. So +1 from me.

File: api/v1alpha1/nodepool_types.go (hunk 2 above)
Comment (enxebre) [reply api/v1alpha1/nodepool_types.go:196]: Right now any change to the template will trigger a rolling upgrade. I think we can probably keep that discussion out of this PR and go through it in https://github.com/openshift/hypershift/pull/486

**Comment by csrwng** [comment]
/approve
We need a folow up to set the HostedCluster tags in the Infrastructure resource that we lay down on the guest cluster
(https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-infrastructure-02-config.yaml)

**Comment by alvaroaleman** [comment]
@csrwng should I create a manual cherrypick for 4.8? The main reason it seems to conflict is that the root volume work wasn't backported (https://github.com/openshift/hypershift/pull/422), should it get backported?

**Comment by csrwng** [comment]
@alvaroaleman Yeah I hadn't backported that one since it didn't impact the control plane operator, but it looks like if we want other changes that impact both sides to apply cleanly, we'll also have to backport them too. You're welcome to backport it and then backport this one. Thank you!

**Pull request merged**